### ISSUE TITLE
ls -F and filename completion run very slowly

### ISSUE THREAD

**Issue by StevePER** [git-for-windows/git#3423]
(Apologies if this is not the right place to open this issue. I am not clear about whether it belongs here, on MSYS2, or on some upstream package.)

 - [X] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.33.0.windows.2
cpu: x86_64
built from commit: 8735530946cced809cc6cc4c2ca3b078cdb3dfc8
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.19042.1165]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
$ cat /etc/install-options.txt

Editor Option: VIM
Custom Editor Path:
Default Branch Option: main
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: LFOnly
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Disabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Disabled
```

 - Any other interesting things about your environment that might be related to the issue you're seeing?

The underlying problem is caused by the Windows Antimalware Service. I am not able to disable or reconfigure this service. The Antimalware Service is scanning files when they are opened by other executables. But why do ls -F and filename completion need to open each file to determine whether they are files or directories?

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Bash

 - What commands did you run to trigger this issue? If you can provide a [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve) this will help us understand the issue.

I have a directory on my local drive containing 3 files, each around 177M in size. I run the following command:
```
$ time find FP4 -type f
FP4/2021-08-18_09-28-58.idc
FP4/2021-08-18_09-30-19.idc
FP4/2021-08-18_09-55-11.idc

real    0m0.072s
user    0m0.000s
sys     0m0.015s

$ time ls -F FP4
2021-08-18_09-28-58.idc  2021-08-18_09-30-19.idc  2021-08-18_09-55-11.idc

real    0m4.177s
user    0m0.000s
sys     0m0.015s
```
 - What did you expect to occur after running these commands?

The ls command should complete in about the same time as the find command takes to run on the directory, given that both commands are checking the file type.

 - What actually happened instead?

Listing the names of the files in the directory takes 4 seconds. The time taken increases with the total size of the files in the directory.

Task Manager shows that "Antimalware Service Executable" is using CPU. Process Explorer shows that the ls process is opening each of the large files in the directory, which explains why Antimalware is scanning them.

The same problem occurs when using <TAB> on the command line for filename completion. In this case, Process Explorer shows that the bash process is opening each of the files.

Why do ls and bash need to open every file to determine if they are files or directories? The find command doesn't need to, and from the output of strace I don't think they do this in Linux.

Running ls (with no parameters) makes it run at the expected speed. Likewise if I remove the "mark-directories" option from inputrc, command line completion runs at the expected speed.

**Comment by dscho** [comment]
I do think you're right in your suspicion that this is an MSYS2 issue. My best guess is that `ls` will have to open the files because it needs to see whether the file is executable (and it could be a script, starting with `#!/bin/sh` or the likes). The fact that `ls` never prints this information is irrelevant because `ls` probably still retrieves information via `lstat()`, and MSYS2's emulation of that POSIX function needs to fill in all the `struct st` fields even if they are not used. If you want to verify whether that is the case, you can use [the `strace` tool](https://cygwin.com/cygwin-ug-net/strace.html) to find out.

**Comment by rimrul** [comment]
> Why do ls and bash need to open every file to determine if they are files or directories? The find command doesn't need to, and from the output of strace I don't think they do this in Linux.

You almost found the answer there yourself at the end: Because `bash` and `ls` are designed for POSIX-ish systems. They rely on [`stat()`](https://linux.die.net/man/2/stat) to check wether a given path is a file or directory. But `stat()` returns more information than that. And since Microsoft doesn't provide a drop-in replacement for `stat()` the `MSYS2` runtime neeeds to provide it's own `stat()` implementation that gathers all that information from multiple Windows API calls. One of those seeems to be considered "opening the file".
The MSYS2 runtime has no way of knowing what part of the information the `stat()` caller needs, so it has to provide all the information.

`find` on the other hand knows exactly what information it needs about a file/directory and can gather just that information.

**Comment by StevePER** [comment]
`strace` shows that `fstat` is taking a very long time.

**Comment by dscho** [comment]
> 
> 
> `strace` shows that `fstat` is taking a very long time.

Yep, that figures. There is unfortunately not much we can do about this...

**Comment by dscho** [comment]
> There is unfortunately not much we can do about this...

Hence I will have to close this ticket without being able to address it.
